docs/rules/no-mutation-props.md

```diff
@@ -0,0 +1,115 @@
+# Prevent mutation of this.props (no-mutation-props)
+
+NEVER mutate `this.props`, as all React components must act like pure functions with respect to their props. 
+Treat `this.props` as if it were immutable. More info available at [props-are-read-only](https://facebook.github.io/react/docs/components-and-props.html#props-are-read-only)
+
+## Rule Details
+
+This rule is aimed to forbid the use of mutating `this.props`.
+
+The following patterns are considered warnings:
+
+```jsx
+var Hello = React.createClass({
+  render: function() {
+    this.props.name = this.props.name.toUpperCase();
+    return <div>Hello {this.props.name}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    const {list} = this.props;
+    list.push(2);
+    return <div>{list.length} things</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    const [firstThing] = this.props.list;
+    firstThing.foo = 'bar';
+    return <div>{firstThing.foo}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    delete this.props.foo;
+    return <div>{Object.keys(this.props).length} props</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    Object.assign(this.props.foo, {bar: 'baz'});
+    return <div>{this.props.foo.bar}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    Object.defineProperty(this.props, 'foo');
+    return <div>{this.props.foo}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    this.props.foo++;
+    return <div>{this.props.foo}</div>;
+  }
+});
+```
+
+## Rule Options
+
+```js
+...
+"react/no-mutation-props": [<enabled>, {
+  "allowArrayMutations": <boolean>,
+  "allowableArrayMutations": <array<string>>,
+  "allowObjectStatics": <boolean>,
+  "allowableObjectStatics": <array<string>>,
+}]
+...
+```
+
+### `allowArrayMutations`
+
+When `true` the rule ignores [array mutation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array#Mutator_methods) of props.
+
+You should only use this option if you have a prop that contains an object with methods that conflict with array mutation methods.
+
+The following patterns are considered okay and do not cause warnings:
+
+```jsx
+this.props.list.push(1)
+```
+
+### `allowablArrayMutations`
+
+When `allowArrayMutation: true` the rule specifies specific [array mutations](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array#Mutator_methods) of props to allow. Mutations not listed will remain as warnings.
+
+You must used this option in conjunction with `allowArrayMutation`. You should only use this option if you have a prop that contains an object with methods that conflict with array mutation methods.
+
+### `allowObjectStatics`
```

Comment:
I don't think this should be restricted to `Object` statics - there's `Reflect` statics that mutate too, and users might provide a single identifier (like `lodash.assign`) that they'd want to be warned on as well.

Reply:
`Reflect` is a good call! I forgot about that. I'll add that in.

Unfortunately, this isn't setup for arbitrary `lodash.assign` calls, but I can see that being a good feature. However, adding support for that is quite a bit of work. I suggest that we look at that in a issue and see if that's something the community wants/expects.

Reply:
That was my intention with allowing the option - I should be able to put any `foo.bar` string in there and have it work. Specifically, I want it to work for `import assign from 'object.assign'; assign()`, and `import lodash from 'lodash'; lodash.assign()`.

Reply:
I gotcha, the issue is that parsing the AST to determine that is more difficult. As I say, I'd love to support that, but it feels like something that can be done on the next iteration of this rule?